src/core/controllers/oidc.go

```diff
@@ -122,30 +123,100 @@ func (oc *OIDCController) Callback() {
-		}
-		u.GroupIDs = gids
-		oidcUser, err := dao.GetOIDCUserByUserID(u.UserID)
-		if err != nil {
-			oc.SendInternalServerError(err)
+		// Recover the username from d.Username by default
+		username := info.Username
+
+		// Fix blanks in username
+		username = strings.Replace(username, " ", "_", -1)
+
+		// If automatic onboard is enabled, skip the onboard page
+		if oidcSettings.AutoOnboard {
+			log.Debug("Doing automatic onboarding\n")
+			user, onboarded := userOnboard(oc, info, username, tokenBytes)
+			if onboarded == false {
+				log.Error("User not onboarded\n")
+				return
+			}
+			log.Debug("User automatically onboarded\n")
+			u = user
+		} else {
+			oc.SetSession(userInfoKey, string(ouDataStr))
+			oc.Controller.Redirect(fmt.Sprintf("/oidc-onboard?username=%s", username), http.StatusFound)
+			// Once redirected, no further actions are done
 			return
 		}
-		_, t, err := secretAndToken(tokenBytes)
-		oidcUser.Token = t
-		if err := dao.UpdateOIDCUser(oidcUser); err != nil {
-			oc.SendInternalServerError(err)
-			return
+	}
+
+	gids, err := group.PopulateGroup(models.UserGroupsFromName(info.Groups, common.OIDCGroupType))
+	if err != nil {
+		log.Warningf("Failed to populate groups, error: %v, user will have empty group list, username: %s", err, info.Username)
+	}
+	u.GroupIDs = gids
+	oidcUser, err := dao.GetOIDCUserByUserID(u.UserID)
+	if err != nil {
+		oc.SendInternalServerError(err)
+		return
+	}
+	_, t, err := secretAndToken(tokenBytes)
+	oidcUser.Token = t
+	if err := dao.UpdateOIDCUser(oidcUser); err != nil {
+		oc.SendInternalServerError(err)
+		return
+	}
+	oc.PopulateUserSession(*u)
+	oc.Controller.Redirect("/", http.StatusFound)
+
+}
+
+func userOnboard(oc *OIDCController, info *oidc.UserInfo, username string, tokenBytes []byte) (*models.User, bool) {
+	s, t, err := secretAndToken(tokenBytes)
+	if err != nil {
+		oc.SendInternalServerError(err)
+		return nil, false
+	}
+
+	gids, err := group.PopulateGroup(models.UserGroupsFromName(info.Groups, common.OIDCGroupType))
+	if err != nil {
+		log.Warningf("Failed to populate group user will have empty group list. username: %s", username)
+	}
+
+	oidcUser := models.OIDCUser{
+		SubIss: info.Subject + info.Issuer,
+		Secret: s,
+		Token:  t,
+	}
+
+	user := models.User{
+		Username:     username,
+		Realname:     username,
+		Email:        info.Email,
+		GroupIDs:     gids,
+		OIDCUserMeta: &oidcUser,
+		Comment:      oidcUserComment,
+	}
+
+	log.Debugf("User created: %+v\n", user)
+
+	err = dao.OnBoardOIDCUser(&user)
+	if err != nil {
+		if strings.Contains(err.Error(), dao.ErrDupUser.Error()) {
+			oc.RenderError(http.StatusConflict, "Conflict, the user with same username or email has been onboarded.")
+			return nil, false
 		}
-		oc.PopulateUserSession(*u)
-		oc.Controller.Redirect("/", http.StatusFound)
+
+		oc.SendInternalServerError(err)
+		oc.DelSession(userInfoKey)
```

Comment:
you have the userinfo removed in the `Onboard`  func so this seems not necessary

Reply:
Do you mean that oc.DelSession(userInfoKey) is already in Onboard func? It is, but only when the onboard is successful, so userOnboard returns (user, true). But if the onboard fails it won't be removed. But I think as this session key is used only for the "manual onboarding" screen, it makes sense to remove that from userOnboard func and remove that key in the Onboard func instead, whether the onboard fails to not. I'll make that change.

Reply:
Fixed now. In fact, it can only be removed when the onboard is successful, so if there are errors and the username is not accepted (i.e. already in use), the information needs to be present in the session.

So now it is only removed in `Onboard` func after successful onboarding.